src/System.Net.Http/src/System/Net/Http/Unix/CurlResponseMessage.cs

```diff
@@ -0,0 +1,247 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.IO;
+using System.Threading;
+
+namespace System.Net.Http
+{
+    internal sealed class CurlResponseMessage : HttpResponseMessage
+    {
+        private readonly CurlResponseStream _responseStream = new CurlResponseStream();
+
+        #region Properties
+
+        internal CurlResponseStream ContentStream
+        {
+            get { return _responseStream; }
+        }
+
+        #endregion
+
+        internal CurlResponseMessage(HttpRequestMessage request)
+        {
+            RequestMessage = request;
+            Content = new StreamContent(_responseStream);
+        }
+    }
+
+        internal sealed class CurlResponseStream : Stream
+        {
+            private volatile bool _disposed = false;
+            private Stream _innerStream;
+            private readonly object _lockObject = new object();
+            private readonly CancellationTokenSource _cancellationTokenSource = new CancellationTokenSource();
+            private readonly ManualResetEventSlim _dataRequestedEvent = new ManualResetEventSlim(false);
+            private readonly ManualResetEventSlim _dataReadyEvent = new ManualResetEventSlim(false);
+            private readonly ManualResetEventSlim _noMoreDataEvent = new ManualResetEventSlim(false);
+
+            internal CurlResponseStream() : base()
+            {
+            }
+
+            internal void SignalComplete(bool isCancellation)
+            {
+                if (!_disposed)
+                {
+                    _cancellationTokenSource.Cancel();
+                }
+            }
+
+            internal unsafe void WaitAndSignalReaders(byte* pointer, long length)
+            {
+                CheckDisposed();
+                _dataRequestedEvent.Wait(Timeout.InfiniteTimeSpan, _cancellationTokenSource.Token);
```

Comment:
Another place it'd be good to add a TODO for follow-up to see if there's any way to eliminate the blocking.  I think you said that when WaitAndSignalReader returns, the byte* will no longer be valid, so we can't do this asynchronously.  We'll want to investigate other options.